Feature/chainkit download

cryptowizard0 opened the pull request:
### Main Changes
- Added download logic for chainkit
- Removed chainkit's dependency on node, replaced with rdb interface
- Added chainkit reference in Node
- Added upload example for uploading all transaction processes of a single pid. Run method: cd examples && go run ./upload [pid]

cryptowizard0 committed:
Implement process transaction download by nonce range

Added DownloadByPid to download all transactions of a process within a nonce range, including spawn transaction verification and batch GraphQL queries. Introduced downloadByNonce, queryByNonce, and related helpers for assignment and message retrieval. Added error handling for invalid message types.

cryptowizard0 committed:
Add message and assignment caching to Chainkit

Introduces Cache and GetCache methods to Chainkit for storing and retrieving message and assignment BundleItems in Redis. Includes comprehensive tests for caching, retrieval, overwriting, concurrent access, and handling of non-existent cache entries.

cryptowizard0 committed:
feat: chainkit download mssage

cryptowizard0 committed:
Refactor cache logic to use txid-based BundleItem storage

Replaces pid/nonce-based cache and retrieval with txid-based BundleItem storage throughout Chainkit and related modules. Updates interface, Redis implementation, and tests to reflect the new cache structure. Improves cache hit logic and simplifies data model for transaction downloads and recovery.

cryptowizard0 committed:
Add IsUploaded check and example upload script

Introduces an IsUploaded method to check for already uploaded transactions and updates the upload logic to skip duplicates. Refactors related DB methods for consistency and adds an example script (examples/upload.go) demonstrating usage of the upload functionality.

cryptowizard0 committed:
Refactor message retrieval in upload example

Removed NodeDB.GetMessageByNonce and updated the upload logic to use db.GetMessageByNonce directly. The nonce range is now determined dynamically using db.GetNonce, improving flexibility and correctness.

cryptowizard0 committed:
upload example : read config from config.yaml

cryptowizard0 committed:
Update recovery.go

cryptowizard0 committed:
clear todo

cryptowizard0 committed:
Refactor Chainkit to use INodeDB and update config

- Replaces INode with INodeDB in Chainkit for improved separation of concerns and updates related method calls.
- Adds nodeRedisUrl to config and updates initialization and usage across cmd, examples, and upload logic. Removes SetChainkit from Node and passes Chainkit directly during Node creation.
- Improves config validation and error handling in cmd/cfgchainkit.go.

cryptowizard0 committed:
Update config.yaml

cryptowizard0 committed:
Add chainkit config and update main.go to use it

Introduced a new config_chainkit.yaml file with chainkit-specific settings and updated main.go to reference this new config file instead of config.yaml.

cryptowizard0 committed:
Update graphql.go

cryptowizard0 committed:
Refactor schema interfaces and move types to node

Moved IDB and INodeDB interfaces from chainkit/schema/types.go to node/schema/idb.go as IDBChainkit and IDBTool, respectively. Deleted types.go and updated references in chainkit.go. Renamed config.go to schema.go and added DownloadResult type to schema.go for better organization.

cryptowizard0 committed:
Merge branch 'develop' into feature/chainkit_download

cryptowizard0 committed:
Relax condition for returning local message and assignment

cryptowizard0 committed:
Move Gin mode and logging setup before config loading

cryptowizard0 committed:
merge from develop

Renamed fields in Result and ResSpawn structs for consistency (Assignments, Sequence). Updated async SpawnAndWait to return result after waiting. Adjusted timestamp handling and error propagation in examples and node logic. Added config_chainkit.yaml and removed chainkit section from config_payment.yaml. Bumped NodeVersion to v0.3.0.

cryptowizard0 committed:
refactor: move chainkit from server to node

Chainkit is now conditionally loaded based on the 'enableChainkit' config flag. Chainkit lifecycle management (Run, Close, AssignmentHandler) has been moved from the server to the node, simplifying the server's responsibilities. Configuration files updated to support the new flag and nodeRedisURL option.

cryptowizard0 committed:
Move IDB interfaces from node/schema to chainkit/schema

The IDBTool and IDBChainkit interfaces were relocated from node/schema/idb.go to chainkit/schema/schema.go for better organization and to centralize schema definitions used by chainkit. References in chainkit/chainkit.go were updated accordingly.

hymatrix-core commented on cmd/cfgchainkit.go:
```diff
@@ -1,17 +1,43 @@
+
+	cfg := chainkitSchema.Config{
+		RedisUrl:     viper.GetString("chainkit.redisURL"),
+		NodeRedisUrl: viper.GetString("chainkit.nodeRedisURL"),
```
Reuse the node’s redis configuration? If the manual configuration is inconsistent with the node’s Redis settings, it may cause exceptions.

hymatrix-core commented on cmd/config_chainkit.yaml:
```diff
@@ -18,6 +18,7 @@ joinNetwork: true
 enableChainkit: true
 chainkit:
   redisURL: redis://@localhost:6379/10
+  nodeRedisURL: redis://@localhost:6379/0
```
remove?

hymatrix-core commented on cmd/config.yaml:
```diff
@@ -15,3 +15,11 @@ nodeURL: http://127.0.0.1:8080
 
 joinNetwork: false
 
+enableChainkit: true
```
Should it be removed from the default configuration items?

hymatrix-core commented on db/rdb/chainkit.go:
```diff
@@ -13,6 +15,7 @@ const (
 	RdbUploadingTxIds   = "chainkit:uploading"            // Set: Uploading TxId pool
 	RdbCurrentBundledIn = "chainkit:current_bundledin_id" // String: current bundledIn id with 1 hour expiration
 	RdbUploadedTxIds    = "chainkit:uploaded_txids"       // Set: Uploaded TxId pool
+	RdbCache            = "chainkit:cache"                // key: TxId, value BundleItem
```
Move all constants to db/rdb/schema/redis.go.

cryptowizard0 replied:
ok

hymatrix-core commented on examples/upload.go:
```diff
@@ -0,0 +1,53 @@
+	goarSchema "github.com/permadao/goar/schema"
+)
+
+type INodeDB interface {
```
This interface definition is invalid in the test case.

cryptowizard0 committed:
Move chainkit Redis constants to schema package

Refactored chainkit Redis key and limit constants from chainkit.go to schema/redis.go for better organization and reuse. Updated all references in chainkit.go and chainkit_test.go to use the schema package constants.

cryptowizard0 committed:
Remove nodeRedisURL from chainkit config

Eliminates the nodeRedisURL field from configuration files and updates code to reference redisURL directly for NodeRedisUrl. This simplifies configuration and reduces redundancy.

cryptowizard0 committed:
Refactor nodeDB initialization in upload example

Simplifies the nodeDB variable initialization by removing the INodeDB interface and directly assigning the result of rdb.New. Cleans up unused imports and interface definitions for clarity.

cryptowizard0 committed:
Remove Chainkit configuration from config.yaml

Deleted Chainkit-related settings including enableChainkit, redisURL, keyfilePath, and optType from the configuration file. This simplifies the config and removes unused or deprecated Chainkit integration.
